src/pgwire/src/protocol.rs

```diff
@@ -2056,6 +2082,68 @@ where
         ))
     }
 
+    async fn copy_to_count(
+        &mut self,
+        rx: oneshot::Receiver<CopyToResponseUnary>,
+    ) -> Result<(State, SendRowsEndedReason), io::Error> {
+        let mut row_count = 0;
+        tokio::select! {
+            e = self.conn.wait_closed() => return Err(e),
+            _ = self.adapter_client.canceled() => {
+                return self
+                    .error(ErrorResponse::error(
+                        SqlState::QUERY_CANCELED,
+                        "canceling statement due to user request",
+                    ))
+                .await.map(|state| (state, SendRowsEndedReason::Canceled));
+            },
+            resp = rx => match resp {
+                Ok(CopyToResponseUnary::Rows(count)) => {
+                    row_count = count;
+                }
+                Ok(CopyToResponseUnary::Error(text)) => {
+                    return self
+                        .error(ErrorResponse::error(SqlState::INTERNAL_ERROR, text.clone()))
+                    .await
+                    .map(|state| (state, SendRowsEndedReason::Errored { error: text }));
+                }
+                Ok(CopyToResponseUnary::Canceled) => {
+                    return self.error(ErrorResponse::error(
+                            SqlState::QUERY_CANCELED,
+                            "canceling statement due to user request",
+                        ))
+                        .await.map(|state| (state, SendRowsEndedReason::Canceled));
+                }
+                Err(_e) => {
+                    // TODO (mouli): how should we handle RecvError
+                    let text = "error receiving response".to_string();
+                    return self
+                    .error(ErrorResponse::error(SqlState::INTERNAL_ERROR, text.clone()))
+                    .await
+                    .map(|state| (state, SendRowsEndedReason::Errored { error: text }));
+                }
+
+            },
+            notice = self.adapter_client.session().recv_notice() => {
+                self.send(notice.into_response())
+                    .await?;
+                self.conn.flush().await?;
+            }
+        }
+
+        self.conn.flush().await?;
+
+        let tag = format!("COPY {}", row_count);
+        self.send(BackendMessage::CopyDone).await?;
+        self.send(BackendMessage::CommandComplete { tag }).await?;
+        Ok((
+            State::Ready,
+            SendRowsEndedReason::Success {
+                rows_returned: row_count,
+            },
+        ))
```

Comment:
cc @mjibson Does this response look okay to you? Also, covered in test [here](https://github.com/MaterializeInc/materialize/pull/25042/files#diff-5224c00f6dd6f3b7c98338d2b3a4973c53c0c2abfe1ed9d1487fb2d5ff1f1b14R22)

Reply:
This is now removed